### PR TITLE
Fix missing or wrong implementations in node.go

### PR DESCRIPTION
**Pull request by leviinz**
Some types in ast/node.go were not implementing the Node type, and some implementations were missing Idx0 and Idx1 functions for their type, which results in errors when attempting to use/import some of the types. I have corrected all the faulty implementations. I believe I have implemented the Idx0 and Idx1 functions correctly, but please review those changes and see if I did anything wrong.

### CODE REVIEW

File: ast/node.go [hunk 1]
```diff
@@ -114,12 +114,12 @@ type (
 
 	DotExpression struct {
 		Left       Expression
-		Identifier Identifier
+		Identifier *Identifier
```
Comment (dop251) [inline ast/node.go:117]: What is the reason for this change (and the other changes to pointer)?

Comment (leviinz) [reply ast/node.go:117]: > What is the reason for this change (and the other changes to pointer)?

The only reason for these changes is because their pointer types inherit the Node interface, but the non-pointer type does not (ex: `*Identifier` implements `Node` because its `Idx0` and `Idx1` implementations have pointer receivers, but `Identifier` is not a `Node`). It's basically just to allow easier access to Node types.

Comment (dop251) [reply ast/node.go:117]: I don't see a problem, you can always use `&dotExpression.Identifier` if you need to pass a Node somewhere, and accessing methods will just work (e.g. `dotExpression.Identifier.Idx1()`). On the other hand, changing the fields to pointers results in extra allocation and increases memory fragmentation.

Comment (leviinz) [reply ast/node.go]: Good point. I've reverted those changes in the latest commit.

File: ast/node.go [hunk 2]
```diff
@@ -721,9 +727,11 @@ func (self *NullLiteral) Idx1() file.Idx        { return file.Idx(int(self.Idx)
 func (self *NumberLiteral) Idx1() file.Idx      { return file.Idx(int(self.Idx) + len(self.Literal)) }
 func (self *ObjectLiteral) Idx1() file.Idx      { return self.RightBrace + 1 }
 func (self *ObjectPattern) Idx1() file.Idx      { return self.RightBrace + 1 }
+func (self *ParameterList) Idx1() file.Idx      { return self.Closing }
```
Comment (dop251) [inline ast/node.go:730]: Should be `self.Closing + 1`, and in expression.go:1179 it should be `Closing: id.Idx1() - 1`

File: ast/node.go [hunk 3]
```diff
@@ -721,9 +727,11 @@ func (self *NullLiteral) Idx1() file.Idx        { return file.Idx(int(self.Idx)
...
 func (self *RegExpLiteral) Idx1() file.Idx      { return file.Idx(int(self.Idx) + len(self.Literal)) }
 func (self *SequenceExpression) Idx1() file.Idx { return self.Sequence[len(self.Sequence)-1].Idx1() }
 func (self *StringLiteral) Idx1() file.Idx      { return file.Idx(int(self.Idx) + len(self.Literal)) }
+func (self *TemplateElement) Idx1() file.Idx    { return self.Idx }
```
Comment (dop251) [inline ast/node.go:734]: Should be `file.Idx(int(self.Idx) + len(self.Literal))`

File: ast/node.go [hunk 4]
```diff
@@ -795,6 +805,10 @@ func (self *PropertyKeyed) Idx1() file.Idx { return self.Value.Idx1() }
 
 func (self *ExpressionBody) Idx1() file.Idx { return self.Expression.Idx1() }
 
+func (self *VariableDeclaration) Idx1() file.Idx {
+	return self.Var
```
Comment (dop251) [inline ast/node.go:809]:
```suggestion
	if len(self.List) > 0 {
		return self.List[len(self.List)-1].Idx1()
	}
	return self.Var + 3
```

**Review by dop251 — CHANGES_REQUESTED** [review]

**Comment by leviinz** [comment]
Fixed the incorrect index implementations mentioned in the review in the latest commit.

**Comment by dop251** [comment]
Thanks!
